### PR TITLE
Add streaming replication

### PR DESCRIPTION
**Pull request by mnagy**
This is based on #13 by @jmccormick2001.

We needed to restructure a lot of stuff to fit in with what we have in MySQL and MongoDB. Opening another PR to make it easier.

@jhadvig did the template and initial restructuring.

Right now it is still a WIP. The problem I have right now is that only the initial replication works, but updates to master will not propagate to slave.

### CODE REVIEW

File: examples/replica/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+This resource defines the `PodTemplate` of the PostgreSQL servers that act as the `slaves` to the `master` server. In the provided JSON example, this Replication Controller starts with 2 slaves. Each `slave` server first waits for the `master` server to become available (getting the `master` server IP using the DNS lookup). Once the `master` is available, the PostgreSQL 'slave' server is started and connected to the `master`. 
+
+To check the 'slave' status, you can issue the following command on the slave container:
+
+```
+psql -h pg-standby.default.local -U user -W db 
+```
+
+You can add more slaves if you want, using the following `osc` command.
```
Comment (jhadvig) [inline examples/replica/README.md]: osc -> oc

File: README.md [hunk 2]
```diff
@@ -144,3 +146,39 @@ User can choose between testing PostgreSQL based on RHEL or CentOS image.
...
+To set up a replication configuration, you would deploy a master Postgres
+instance first, then create any number of read-only replica instances, all of 
+which would point to the master.
+
+The environment variable POSTGRESQL_MASTER is passed to the replica upon
+startup which will be used to perform the connection from the replica to
+the master.  This value can be an IP address or a host name if you are
+running within a Docker DNS orchestration environment (e.g. Openshift).
+
+The environment variable POSTGRESQL_REPLICA when set, is used to denote
```
Comment (jhadvig) [inline README.md]: remove this paragraph

File: examples/replica/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+```
+$ osc update rc pg-slave --patch='{ "apiVersion": "v1beta1", "desiredState": { "replicas": 3 }}'
```
Comment (jhadvig) [inline examples/replica/README.md]: osc update -> oc scale

File: README.md [hunk 4]
```diff
@@ -65,6 +65,8 @@ initialization, by passing `-e VAR=VALUE` to the Docker run command.
 |  `POSTGRESQL_PASSWORD`       | Password for the user account                  |
 |  `POSTGRESQL_DATABASE`       | Database name                                  |
 |  `POSTGRESQL_ADMIN_PASSWORD` | Password for the `postgres` admin account (optional)     |
+|  `POSTGRESQL_MASTER`         | IP or hostname of the `postgres` master, used by replicas (optional)     |
+|  `POSTGRESQL_REPLICA`        | yes or no specifies if we are configuring a `postgres` replica (optional)     |
```
Comment (jhadvig) [inline README.md]: We dont need this two envvars

Comment (jhadvig) [inline README.md]: We dont need this two envvars

File: 9.2/contrib/common.sh [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  while [ true ]; do
+    master_addr=$(postgresql_master_addr)
+    [ ! -z "${master_addr}" ] && break
+    echo "Waiting for PostgreSQL master service ..."
+    sleep 1
+  done
+  echo "Got PostgreSQL master service address: ${master_addr}"
+  while [ true ]; do
+    # TODO port to PostgreSQL
+    # mysqladmin --host=${master_addr} --user="${MYSQL_MASTER_USER}" \
```
Comment (jhadvig) [inline 9.2/contrib/common.sh]: we need to create a response check similar to this `postgresql ping`

File: 9.2/contrib/common.sh [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  done
+}
+
+# postgresql_master_addr lookups the 'postgresql-master' DNS and get list of the available
+# endpoints. Each endpoint is a PostgreSQL container with the 'master' PostgreSQL running.
+function postgresql_master_addr() {
+  local service_name=${POSTGRESQL_MASTER_SERVICE_NAME:-postgresql-master}
+  local endpoints=$(dig ${service_name} A +search +short 2>/dev/null)
+  # FIXME: This is for debugging (docker run)
+  if [ -v POSTGRESQL_MASTER_IP ]; then
```
Comment (jhadvig) [inline 9.2/contrib/common.sh]: where is this envvar set ?

File: 9.2/contrib/common.sh [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    endpoints=${POSTGRESQL_MASTER_IP-}
```
Comment (jhadvig) [inline 9.2/contrib/common.sh]: `${POSTGRESQL_MASTER_IP:-}`

File: examples/replica/postgres_replica.json [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+      "from": "[a-zA-Z0-9]{12}"
+    },
+    {
+      "name": "POSTGRESQL_DATABASE",
+      "description": "The name of the database that will be created",
+      "value": "userdb"
+    },
+    {
+      "name": "POSTGRESQL_ADMIN_PASSWORD",
+      "description": "The password for the PostgreSQL adminitrator",
```
Comment (tristanz) [inline examples/replica/postgres_replica.json]: typo "administrator"

**Comment by mfojtik** [comment]
@mnagy ping, any progress on this?

File: 9.2/contrib/common.sh [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  fi
+  echo -n "$(echo $endpoints | cut -d ' ' -f 1)"
```
Comment (mfojtik) [inline 9.2/contrib/common.sh]: this means you get just the first master IP correct? if true, then you have to document it.

File: 9.2/run-postgresql-slave.sh [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+set -eu
+
+source ${HOME}/common.sh
+
+function initialize_replica() {
+  echo "Initializing replica"
```
Comment (mfojtik) [inline 9.2/run-postgresql-slave.sh]: add " ..." suffix

File: 9.2/run-postgresql-slave.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+EOF
+}
+
+check_env_vars
+generate_postgresql_config
+generate_passwd_file
+initialize_replica
+unset_env_vars
+
+# Testing
```
Comment (mfojtik) [inline 9.2/run-postgresql-slave.sh]: what are you testing here?

File: README.md [hunk 12]
```diff
@@ -144,3 +146,39 @@ User can choose between testing PostgreSQL based on RHEL or CentOS image.
...
+For example, a recovery.conf file is created and placed in PGDAT when this
+variable is set.
+
+The master or regular standalone Postgres instances are pre-configured to
+allow replication from any host using the user/password that is created
+when the master or regular standalone Postgres instance is created.
+
+Here is an example of creating a replica instance and mapping it's Postgres port to localhost port 5433, it connects to a Postgres master at 172.17.0.5:
+```
+$ docker run -d --name postgresql_database -e POSTGRESQL_MASTER=172.17.0.5 -e POSTGRESQL_REPLICA=yes -e POSTGRESQL_USER=user -e POSTGRESQL_PASSWORD=pass -e POSTGRESQL_DATABASE=db -p 5433:5432 openshift/postgresql-92-centos7
```
Comment (mfojtik) [inline README.md]: What you want is description of how the OpenShift/K8S template does this for you... We don't expect users writing this crazy command lines

File: examples/replica/README.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+### Service 'pg-slave'
+
+This provides a Service for the PostgreSQL servers that the master uses as 'slaves' which are used to replicate the data from the PostgreSQL master.
+
+In this case, you can query the DNS (eg. `dig pg-standby A +search +short`) to
+obtain the list of the Service endpoints (the PostgreSQL servers that subscribe to this service).
+
+### ReplicationController 'pg-master'
+
+This resource defines the `PodTemplate` of the PostgreSQL server that acts as the 'master'. The Pod uses the `openshift/postgresql-92-centos7` image.  Once the PostgreSQL master server is started, it has no slaves preconfigured as the slaves registers automatically.  To check that the master PostgreSQL server is working, you can issue the following command on the master container:
```
Comment (mfojtik) [inline examples/replica/README.md]: Can I have multi-master? If not, we should say here that you can have just one master.

File: 9.2/contrib/common.sh (hunk 6 above)
Comment (mnagy) [inline 9.2/contrib/common.sh]: docker run

**Comment by mnagy** [comment]
@mfojtik got sidetracked. I'm back on it :)

File: 9.2/run-postgresql-slave.sh (hunk 11 above)
Comment (mnagy) [inline 9.2/run-postgresql-slave.sh]: I meant to test whether postgres can boot up correctly. Before that, there was something else that . Just realized it's not really needed now, will nuke it.

File: README.md (hunk 12 above)
Comment (mnagy) [inline README.md]: I think we should want that in OS docs, not in this readme. And we want it in examples/replica/README.md. I'm now seeing that it's outdated, it mentions Config object. Will need to go back to that and rewrite.

**Comment by mnagy** [comment]
OK, so I've *finally* discovered where the problem is. The reason why replication wasn't working was that postgres was connecting without a password. This means we need to somehow keep the user password in the container. I don't like that. We should work out a proper way, i.e. creating a special replication user and generating password for him. I don't think we can do the generation inside of the container, so it needs to be provided by an environment variable. It'll then be saved into the recovery.conf file in the conninfo param. This will require a lot of changes to happen, I'll update the PR on Tuesday.

**Comment by mnagy** [comment]
Added various fixes, the replication now works like a charm with OS. PTAL

File: 9.2/contrib/common.sh [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+  echo "  POSTGRESQL_SHARED_BUFFERS (default: 32MB)"
+  exit 1
+}
+
+function check_env_vars() {
+  if ! [[ -v POSTGRESQL_USER && -v POSTGRESQL_PASSWORD && -v POSTGRESQL_DATABASE ]]; then
+    usage
+  fi
+
+  [[ "$POSTGRESQL_USER"     =~ $psql_identifier_regex ]] || usage
```
Comment (bparees) [inline 9.2/contrib/common.sh]: spacing

File: 9.2/contrib/common.sh [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+  unset POSTGRESQL_PASSWORD
+  unset POSTGRESQL_DATABASE
+  unset POSTGRESQL_ADMIN_PASSWORD
+}
+
+# postgresql_master_addr lookups the 'postgresql-master' DNS and get list of the available
+# endpoints. Each endpoint is a PostgreSQL container with the 'master' PostgreSQL running.
+function postgresql_master_addr() {
+  local service_name=${POSTGRESQL_MASTER_SERVICE_NAME:-postgresql-master}
+  local endpoints=$(dig ${service_name} A +search | grep ";${service_name}" | cut -d ';' -f 2 2>/dev/null)
```
Comment (bparees) [inline 9.2/contrib/common.sh]: i take it you resolved the DNS lookup issues we were discussing?

File: 9.2/contrib/openshift-custom-recovery.conf.template [hunk 16]
```diff
@@ -0,0 +1,9 @@
+#
+# Custom OpenShift configuration.
+#
+# NOTE: This file is rewritten everytime the container is started!
+#       Changes to this file will be overrwritten.
```
Comment (bparees) [inline 9.2/contrib/openshift-custom-recovery.conf.template]: s/overrwritten/overwritten/

File: examples/replica/README.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
+# PostgreSQL Replication Example
+
+**WARNING:**
+
+**This is only a Proof-Of-Concept example and it is not meant to be used in any
```
Comment (bparees) [inline examples/replica/README.md]: s/any//

File: examples/replica/README.md [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+Replication enables data from one PostgreSQL database server (the master) to be replicated to one or more PostgreSQL database servers (the slaves). Replication is asynchronous - slaves do not need not be connected permanently to receive updates from the master. This means that updates can occur over long-distance connections and even over temporary or intermittent connections such as a dial-up service.  Depending on the configuration, you can replicate all databases, selected databases, or even selected tables within a database.
+
+
+## How does this example work?
+
+The provided JSON file (`postgresql_replica.json`) contains a `DeploymentConfig` resource that groups the Kubernetes and OpenShift resources which are meant to be created.
+
+### Service 'postgresql-master'
+
+This resource provides Service for the PostgreSQL server which acts as the 'master'.
```
Comment (bparees) [inline examples/replica/README.md]: a Service

File: 9.2/contrib/common.sh (hunk 15 above)
Comment (mnagy) [inline 9.2/contrib/common.sh]: Yes. The dig with grep will produce FQDN. This is still a hack, though.

**Comment by bparees** [comment]
couple nits and needs a squash and then lgtm.

**Comment by mnagy** [comment]
Addressed all comments. I don't want to squash the commits, because they were authored by a three different people.

PTAL

**Comment by bparees** [comment]
lgtm.

**Pull request merged**